src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs

```diff
@@ -659,8 +683,29 @@ private List<ODataPathSegment> ProcessSelectTokenPath(PathSegmentToken tokenIn)
             Debug.Assert(tokenIn != null, "tokenIn != null");
 
             List<ODataPathSegment> pathSoFar = new List<ODataPathSegment>();
+
             IEdmStructuredType currentLevelType = this.edmType;
 
+            // If we have a previous level binding with the last segment as a Type segment,
+            // The currentLevelEntityType should be the Type of the TypeSegment
+            // E.g /Orders?$expand=Customer/NS.VipCustomer($select=VipCustomerName)
+            // If we are Binding the selected property VipCustomerName, currentLevelEntityType should be the type of the VipCustomer
+            if (this.parsedSegments.Count > 0 && this.parsedSegments.Last() is TypeSegment)

```

Comment:
I have all of the same micro-optimization feedback here, though I do wonder if we can factor this code into a common method so it doesn't need to be duplicated

Reply:
#Resolved